### ISSUE TITLE
It's possible for desktop to get stuck in an infinite loop of checking out between two branches

### ISSUE THREAD

**Issue by jdngray77** [desktop/desktop#11130]
### Describe the bug

With unknown conditions, switching to a specific branch in my repo causes Desktop to infinitely checkout, flipping between the current and selected branch

### Version & OS
Tested with 100% reproducibility on

2.6.0 on
MacOS 10.12.6,
MacOS 10.15.6,

2.5.7 on MacOS 11.0
2.5.4 on Windows 10 Pro 10.0.19041

### Steps to reproduce the behavior

(Failed to acquire the conditions that cause this for preliminary steps)

1. Clone the [preservation fork](https://github.com/Pheonix-org/Object-Oriented-Design-Portfolio)
2. From the branches drop down, select the `broken_reimport` or `BugPreservation` branch to checkout

### Expected behaviour

The selected branch is checked out.

### Actual behavior

The selected branch is checked out, then your original branch, then back and forth infinitely.

i.e `master > broken_reimport > master > broken_reimport > master > broken_reimport ...`

This continues until the application is force quit.

### Recording

[Infinite switching](https://youtu.be/VAjllL_X1Js)

### Additional context

Whilst I've struggled to find the exact conditions that actually cause this, I believe the state of my specific repository is the key here.

There's a few clues I'd like to list
1. Any fresh clone of the repo comes with a staged but [uncommitted change on the master.](https://www.youtube.com/watch?v=8Ootg6tWKkE)
2. The change is in a package that was recently refactored:
`com.shinkson47.portfolio.labs.six.Aggregation` > `com.shinkson47.portfolio.labs.six.aggregation`

3. Whilst the checkout loop is happening, the staged change flips between:
 > Aggregation ; lines 2-5 added
 > aggregation ; lines 2-5 removed

4. The same above behaviour can also be seen when performing a manual '[Discard Changes](https://youtu.be/OF0mSJQX-o4)'

5. It only happens with the `broken_reimport` branch, or a branch based on it.

6. It occurs regardless of the state of the stash, or if I choose to bring the stray staged change that I mentioned before with us, or leave it behind.

### All links
[Repository preservation fork](https://github.com/Pheonix-org/Object-Oriented-Design-Portfolio)
This repo is a mess, I'm aware.

[Infinite switching recording](https://youtu.be/VAjllL_X1Js)
[Uncommitted change on fresh clone recording](https://www.youtube.com/watch?v=8Ootg6tWKkE)
[Strange discard changes behaviour](https://youtu.be/OF0mSJQX-o4)

**Comment by niik** [comment]
Hi @Shinkson47, sorry to hear you're having problems with GitHub Desktop. This specific error was first reported in https://github.com/desktop/desktop/issues/9297 and was fixed in #10957 which shipped to our [beta](https://github.com/desktop/desktop/blob/development/README.md#beta-channel) channel yesterday and will make its way to our production channel within a couple of weeks.

If you don't want to wait for the production release you can install the beta channel right now and help us test the fix. In this particular case though the beta channel will stop the infinite loop but it won't let you check out that other branch (and neither would Git on the command line) since it looks like the source of the problem is that you've got several paths that only differ by case

![image](https://user-images.githubusercontent.com/634063/100357415-f2fb1900-2ff4-11eb-9ac7-863bd5fe7d1c.png)

```
Markuss-MacBook-Pro:Object-Oriented-Design-Portfolio markus$ git ls-tree -r HEAD --name-only |grep six| grep PlayListApp.java
src/com/shinkson47/portfolio/labs/six/Aggregation/main/PlayListApp.java
src/com/shinkson47/portfolio/labs/six/aggregation/main/PlayListApp.java
```

Those files are what's causing problems here because one of them has changed that the other one doesn't but you've got more files than that which only differ by case:

```
Markuss-MacBook-Pro:Object-Oriented-Design-Portfolio markus$ git ls-tree -r HEAD --name-only | tr '[:upper:]' '[:lower:]' | sort | uniq -c | sort -nr | grep -vE '\s+1'
   2 src/com/shinkson47/portfolio/labs/six/composition/main/rollabledemo.java
   2 src/com/shinkson47/portfolio/labs/six/composition/main/employeedemo.java
   2 src/com/shinkson47/portfolio/labs/six/composition/lib/employee/name.java
   2 src/com/shinkson47/portfolio/labs/six/composition/lib/employee/employee.java
   2 src/com/shinkson47/portfolio/labs/six/composition/lib/employee/date.java
   2 src/com/shinkson47/portfolio/labs/six/composition/lib/dice/rollable.java
   2 src/com/shinkson47/portfolio/labs/six/composition/lib/dice/pairofdice.java
   2 src/com/shinkson47/portfolio/labs/six/composition/lib/dice/die.java
   2 src/com/shinkson47/portfolio/labs/six/aggregation/main/playlistapp.java
   2 src/com/shinkson47/portfolio/labs/six/aggregation/main/employeeregisterdemo.java
   2 src/com/shinkson47/portfolio/labs/six/aggregation/lib/playlist/song.java
   2 src/com/shinkson47/portfolio/labs/six/aggregation/lib/playlist/playlist.java
   2 src/com/shinkson47/portfolio/labs/six/aggregation/lib/employeeregister/name.java
   2 src/com/shinkson47/portfolio/labs/six/aggregation/lib/employeeregister/employeeregister.java
   2 src/com/shinkson47/portfolio/labs/six/aggregation/lib/employeeregister/employee.java
   2 src/com/shinkson47/portfolio/labs/six/aggregation/lib/employeeregister/date.java
```

Both macOS and Windows rely on case-insensitive yet case-preserving file systems which means that those two paths (while legitimate in Git and possible to use individually on Linux) are treated as the same file by the file system which is tripping Git up. See https://docs.microsoft.com/en-us/azure/devops/repos/git/case-sensitivity?view=azure-devops for a great introduction to this problem.

Since the issue with the infinite loop has been solved already and the secondary issue of paths that only differ in case is a hard limitation of the file systems on the platforms we support I'm going to close this. I would recommend either using the GitHub.com web interface to delete one of the missing files

**Comment by jdngray77** [comment]
Thanks @niik for helping me understand that case conflicts in Git are a bigger issue than I ever knew.

I'll get around to sorting this messy repo and manually fixing the conflicts.

Glad to see that the issue has already been handled and in the beta channel, although had I searched the repo before I could've learnt this and found the many existing issues for the same problem and ultimately saved an issue. Learning point for the future!
